File: aulas-1sem/labs/aula14-lab/resgate/portal/pkce.ts
/**
 * Authorization Code + PKCE no navegador.
 *
 * RESGATE: esta é a versão COM a lacuna TODO-5 preenchida.
 *
 * Use quando travar, e não como atalho:
 *     cp resgate/portal/pkce.ts portal/src/auth/pkce.ts
 * Registre `USEI_O_RESGATE: sim` em `docs/EVIDENCIAS.md` e siga.
 *
 * No Passo 3 você fez este fluxo à mão, com o `pkce.py`, e viu cada
 * parâmetro. Aqui ele vira código de produção: as mesmas quatro peças, em
 * TypeScript, rodando dentro do navegador.
 *
 * Duas coisas que este arquivo NÃO faz, e não é esquecimento:
 *
 *   1. não guarda `client_secret`. O portal é um client público. Qualquer
 *      segredo embutido numa SPA está visível na aba de rede, o que faz dele
 *      exatamente o contrário de um segredo. É esse buraco que o PKCE fecha;
 *
 *   2. não valida a assinatura do token. Quem valida é o backend, contra o
 *      JWKS. O navegador só carrega o token de um lado para o outro.
 *
 * Confira sem subir nada:
 *     cd portal && npm test
 */

/** O endereço pelo qual o NAVEGADOR alcança o Keycloak. */
const KEYCLOAK: string =
  import.meta.env.VITE_KEYCLOAK_URL ?? 'http://localhost:8090';
const REALM = 'logitech';
const CLIENT_ID: string =
  import.meta.env.VITE_OIDC_CLIENT_ID ?? 'logitech-portal';

export const AUTORIZACAO = `${KEYCLOAK}/realms/${REALM}/protocol/openid-connect/auth`;
export const TOKEN = `${KEYCLOAK}/realms/${REALM}/protocol/openid-connect/token`;
export const SAIDA = `${KEYCLOAK}/realms/${REALM}/protocol/openid-connect/logout`;

/** A URI de retorno precisa estar registrada no realm, byte a byte. */
export const RETORNO = `${window.location.origin}/`;

export interface Sessao {
  accessToken: string;
  expiraEm: number; // epoch em segundos
  usuario: string;
  papeis: string[];
}

// ---------------------------------------------------------------------------
// Ferramentas: CONGELADAS, já funcionam
// ---------------------------------------------------------------------------

/**
 * base64url sem o preenchimento `=`, como manda a RFC 7636.
 *
 * `btoa` devolve base64 comum, com `+`, `/` e `=`. Nenhum dos três sobrevive
 * a uma query string sem escape, e é por isso que o padrão exige a variante
 * url-safe. Trocar os três caracteres é literalmente toda a diferença.
 */
export function base64url(bytes: Uint8Array): string {
  let bruto = '';
  bytes.forEach((b) => {
    bruto += String.fromCharCode(b);
  });
  return btoa(bruto).replace(/\+/g, '-').replace(/\//g, '_').replace(/=+$/, '');
}

/** Decodifica uma das três partes de um JWT. Não valida nada. */
export function decodificarParte(parte: string): Record<string, unknown> {
  const normal = parte.replace(/-/g, '+').replace(/_/g, '/');
  const preenchido = normal + '='.repeat((4 - (normal.length % 4)) % 4);
  return JSON.parse(decodeURIComponent(escape(atob(preenchido))));
}

// ---------------------------------------------------------------------------
// TODO-5a: o par verificador/desafio
// ---------------------------------------------------------------------------

/**
 * O `code_verifier` (era o TODO-5a).
 *
 * Contrato (RFC 7636, seção 4.1):
 *   - aleatório criptograficamente seguro, e não `Math.random()`;
 *   - entre 43 e 128 caracteres do alfabeto url-safe.
 *
 * 48 bytes viram 64 caracteres, que está dentro da faixa.
 */
export function gerarVerifier(): string {
  return base64url(crypto.getRandomValues(new Uint8Array(48)));
}

/**
 * O `code_challenge` (era o TODO-5b).
 *
 * Contrato: `base64url(SHA-256(verifier))`, que é o método `S256`.
 *
 * Existe também o método `plain`, em que o desafio é o próprio verificador.
 * Ele está no padrão por compatibilidade e não serve para nada aqui: quem
 * interceptar o desafio na URL fica com o verificador de graça. O realm da
 * LogiTech exige `S256`, e o Keycloak recusa `plain`.
 *
 */
export async function desafioS256(verifier: string): Promise<string> {
  const resumo = await crypto.subtle.digest(
    'SHA-256',
    new TextEncoder().encode(verifier),
  );
  return base64url(new Uint8Array(resumo));
}

// ---------------------------------------------------------------------------
// TODO-5c: a ida
// ---------------------------------------------------------------------------

/**
 * A URL de autorização (era o TODO-5c).
 *
 * Parâmetros exigidos, todos na query string:
 *   client_id, response_type=code, scope=openid profile email,
 *   redirect_uri, state, code_challenge, code_challenge_method=S256
 *
 * `URLSearchParams` e não concatenação: o `redirect_uri` tem `:` e `/` e
 * precisa ser escapado.
 *
 * O `state` não é decoração. Ele é o que permite reconhecer, na volta, que
 * a resposta pertence à requisição que ESTE navegador iniciou. Sem ele, um
 * terceiro consegue empurrar um código de autorização dele para a sua
 * sessão, e você termina logado na conta do atacante sem perceber.
 */
export function urlDeAutorizacao(desafio: string, estado: string): string {
  const parametros = new URLSearchParams({
    client_id: CLIENT_ID,
    response_type: 'code',
    scope: 'openid profile email',
    redirect_uri: RETORNO,
    state: estado,
    code_challenge: desafio,
    code_challenge_method: 'S256',
  });
  return `${AUTORIZACAO}?${parametros.toString()}`;
}

// ---------------------------------------------------------------------------
// TODO-5d: a volta
// ---------------------------------------------------------------------------

/**
 * A troca do código pelo token (era o TODO-5d).
 *
 * `POST` para `TOKEN`, corpo `application/x-www-form-urlencoded` com:
 *   grant_type=authorization_code, client_id, code, redirect_uri, code_verifier
 *
 * Repare no que NÃO vai: senha e segredo de cliente. O que prova que quem
 * está trocando o código é quem o pediu é o `code_verifier`, e ele nunca
 * passou pela URL.
 *
 * Devolva a resposta JSON do Keycloak (`access_token`, `expires_in`, ...).
 * Em caso de erro, lance `new Error` com o corpo devolvido: a mensagem do
 * Keycloak diz exatamente o que faltou.
 */
export async function trocarCodigoPorToken(
  codigo: string,
  verifier: string,
): Promise<{ access_token: string; expires_in: number }> {
  const corpo = new URLSearchParams({
    grant_type: 'authorization_code',
    client_id: CLIENT_ID,
    code: codigo,
    redirect_uri: RETORNO,
    code_verifier: verifier,
  });

  const resposta = await fetch(TOKEN, {
    method: 'POST',
    headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
    body: corpo.toString(),
  });

  if (!resposta.ok) {
    const detalhe = await resposta.text();
    throw new Error(`o Keycloak recusou a troca (${resposta.status}): ${detalhe}`);
  }
  return (await resposta.json()) as { access_token: string; expires_in: number };
}

// ---------------------------------------------------------------------------
// Montagem da sessão: CONGELADA
// ---------------------------------------------------------------------------

/**
 * Lê do token o que a interface precisa mostrar.
 *
 * Os papéis vêm de `realm_access.roles`, o MESMO lugar de onde o serviço
 * Java e o serviço Node leem. Se a tela lesse de outro lugar, o botão
 * apareceria para quem o backend vai recusar, e o usuário levaria um 403 de
 * um botão que a própria aplicação ofereceu.
 *
 * Isto NÃO é segurança: é conveniência de interface. Esconder um botão não
 * protege rota nenhuma, porque qualquer um monta a requisição à mão. A
 * segurança de verdade acontece nos dois serviços de backend.
 */
export function sessaoDoToken(accessToken: string, ): Sessao {
  const conteudo = decodificarParte(accessToken.split('.')[1]) as {
    preferred_username?: string;
    exp: number;
    realm_access?: { roles?: string[] };
  };
  return {
    accessToken,
    expiraEm: conteudo.exp,
    usuario: conteudo.preferred_username ?? '(sem nome)',
    papeis: conteudo.realm_access?.roles ?? [],
  };
}

File: aulas-1sem/labs/aula14-lab/pkce.py
#!/usr/bin/env python3
"""O fluxo Authorization Code + PKCE, feito a mao, pelo seu navegador.

CONGELADO: não é tarefa do laboratório. É o instrumento do Passo 3.

Este programa não esconde nada. Ele imprime o `code_verifier`, o
`code_challenge`, a URL de autorização inteira com todos os parâmetros, o
código que voltou e o token decodificado em três partes. Você faz o login
**no navegador**, como um usuário faria, e vê cada peça do fluxo aparecer.

Só depois disto o roteiro libera `curl` com token colado. A ordem é
deliberada: quem começa pelo `curl` decora um comando; quem começa pelo
navegador entende por que o comando funciona.

    python3 pkce.py                      # entra como ana.cliente
    python3 pkce.py --usuario carla.admin
    python3 pkce.py --so-token           # imprime só o access_token, para
                                         # usar em: TOKEN=$(python3 pkce.py --so-token)

Sem dependência nenhuma além da biblioteca padrão.
"""
import argparse
import base64
import hashlib
import http.server
import json
import os
import secrets
import sys
import threading
import urllib.parse
import urllib.request
import webbrowser

# O endereço pelo qual o SEU NAVEGADOR alcança o Keycloak. Note que não é o
# mesmo pelo qual os serviços o alcançam de dentro da rede do Compose
# (`http://keycloak:8090`), e essa diferença é o assunto do Passo 4.
KEYCLOAK = os.environ.get("LOGITECH_KEYCLOAK_URL", "http://localhost:8090")
REALM = "logitech"
CLIENT_ID = "logitech-portal"

# A porta 5199, e não a 5173, para este exercício não disputar o servidor de
# desenvolvimento do portal. As duas estão registradas no realm: URI de
# retorno precisa ser cadastrada ANTES, e é isso que impede um atacante de
# mandar o código de autorização para o servidor dele.
PORTA_RETORNO = 5199
REDIRECT_URI = "http://localhost:%d/callback" % PORTA_RETORNO


def b64url(dados: bytes) -> str:
    """base64url sem o preenchimento `=`, como manda a RFC 7636."""
    return base64.urlsafe_b64encode(dados).decode("ascii").rstrip("=")


def decodificar(parte: str) -> dict:
    return json.loads(base64.urlsafe_b64decode(parte + "=" * (-len(parte) % 4)))


class Retorno(http.server.BaseHTTPRequestHandler):
    """Servidor de uma requisição só: recebe o redirecionamento do Keycloak."""

    recebido = {}

    def do_GET(self):  # noqa: N802
        Retorno.recebido = dict(urllib.parse.parse_qsl(
            urllib.parse.urlparse(self.path).query))
        corpo = ("<html><meta charset='utf-8'><body style='font-family:sans-serif;"
                 "padding:40px'><h2>Pode fechar esta aba.</h2>"
                 "<p>O codigo de autorizacao voltou para o terminal.</p>"
                 "</body></html>").encode("utf-8")
        self.send_response(200)
        self.send_header("Content-Type", "text/html; charset=utf-8")
        self.send_header("Content-Length", str(len(corpo)))
        self.end_headers()
        self.wfile.write(corpo)

    def log_message(self, *args):
        pass


def esperar_codigo(estado: str, segundos: int = 180) -> str:
    try:
        servidor = http.server.HTTPServer(("127.0.0.1", PORTA_RETORNO), Retorno)
    except OSError:
        raise SystemExit(
            "A porta %d ja esta ocupada. Quase sempre e uma execucao anterior deste\n"
            "mesmo programa que ficou esperando um login que nunca chegou.\n"
            "Encerre com:  lsof -ti tcp:%d | xargs kill"
            % (PORTA_RETORNO, PORTA_RETORNO))
    servidor.timeout = segundos
    t = threading.Thread(target=servidor.handle_request, daemon=True)
    t.start()
    t.join(segundos)
    servidor.server_close()

    dados = Retorno.recebido
    if not dados:
        raise SystemExit("Nada voltou em %ds. O login foi concluido no navegador?" % segundos)
    if "error" in dados:
        raise SystemExit("O Keycloak recusou: %s - %s"
                         % (dados.get("error"), dados.get("error_description", "")))
    if dados.get("state") != estado:
        # O `state` existe para isto: casar a resposta com a requisição que
        # você mesmo iniciou. Sem ele, um terceiro consegue empurrar um
        # código de autorização dele para a sua sessão (CSRF de login).
        raise SystemExit("state divergente: esperava %r, veio %r"
                         % (estado, dados.get("state")))
    return dados["code"]


def principal():
    p = argparse.ArgumentParser(description="Fluxo PKCE a mao, pelo navegador.")
    p.add_argument("--usuario", default="ana.cliente",
                   help="so para lembrar voce de quem entrar (o login e no navegador)")
    p.add_argument("--so-token", action="store_true",
                   help="imprime apenas o access_token, sem explicacao")
    p.add_argument("--sem-navegador", action="store_true",
                   help="nao tenta abrir o navegador, so imprime a URL")
    args = p.parse_args()
    falar = not args.so_token

    def diga(*a):
        if falar:
            print(*a)

    # -------------------------------------------------------------------
    # 1. O segredo que nunca sai da sua maquina
    # -------------------------------------------------------------------
    verifier = b64url(secrets.token_bytes(48))          # 64 caracteres
    challenge = b64url(hashlib.sha256(verifier.encode("ascii")).digest())
    estado = b64url(secrets.token_bytes(12))

    diga("\n1) PKCE: o par que so voce conhece")
    diga("   code_verifier  (%d caracteres, FICA AQUI, nunca vai pela rede agora)" % len(verifier))
    diga("       %s" % verifier)
    diga("   code_challenge = base64url(SHA-256(verifier))  ->  vai na URL")
    diga("       %s" % challenge)
    diga("   state          = %s" % estado)
    diga("\n   O navegador leva o DESAFIO. O verificador so aparece no ultimo passo,")
    diga("   direto do seu processo para o Keycloak. Quem interceptar o codigo de")
    diga("   autorizacao no meio do caminho nao consegue troca-lo por token, porque")
    diga("   nao tem o verificador. E para isso que o PKCE existe.")

    # -------------------------------------------------------------------
    # 2. A URL de autorizacao
    # -------------------------------------------------------------------
    parametros = {
        "client_id": CLIENT_ID,
        "response_type": "code",
        "scope": "openid profile email",
        "redirect_uri": REDIRECT_URI,
        "state": estado,
        "code_challenge": challenge,
        "code_challenge_method": "S256",
    }
    url = "%s/realms/%s/protocol/openid-connect/auth?%s" % (
        KEYCLOAK, REALM, urllib.parse.urlencode(parametros))

    diga("\n2) Abra esta URL no navegador e entre como %s (senha: logitech)" % args.usuario)
    diga("   %s\n" % url)
    diga("   Repare: nao ha senha nenhuma nesta URL, e nao ha segredo de cliente.")
    diga("   O portal e um client PUBLICO: qualquer segredo embutido nele estaria")
    diga("   visivel na aba de rede do navegador.\n")

    if not args.sem_navegador:
        try:
            webbrowser.open(url)
        except Exception:
            pass

    codigo = esperar_codigo(estado)
    diga("3) O Keycloak redirecionou para %s com:" % REDIRECT_URI)
    diga("   code  = %s" % codigo)
    diga("   state = confere com o que enviamos\n")
    diga("   Este codigo vale UMA vez e por poucos segundos. Sozinho ele nao serve")
    diga("   para nada: falta o verificador.\n")

    # -------------------------------------------------------------------
    # 3. A troca
    # -------------------------------------------------------------------
    corpo = urllib.parse.urlencode({
        "grant_type": "authorization_code",
        "client_id": CLIENT_ID,
        "code": codigo,
        "redirect_uri": REDIRECT_URI,
        "code_verifier": verifier,
    }).encode("utf-8")

    endereco = "%s/realms/%s/protocol/openid-connect/token" % (KEYCLOAK, REALM)
    with urllib.request.urlopen(endereco, data=corpo, timeout=15) as r:
        resposta = json.loads(r.read())

    acesso = resposta["access_token"]

    if args.so_token:
        print(acesso)
        return

    cabecalho, conteudo, assinatura = acesso.split(".")
    print("4) Troca do codigo pelo token, com o code_verifier junto")
    print("   POST %s" % endereco)
    print("   expires_in = %s segundos" % resposta["expires_in"])
    print("   o access_token tem %d caracteres, em tres partes\n" % len(acesso))

    print("   CABECALHO")
    print("   " + json.dumps(decodificar(cabecalho), indent=2, ensure_ascii=False).replace("\n", "\n   "))

    dados = decodificar(conteudo)
    print("\n   CONTEUDO (os campos que importam hoje)")
    for campo in ("iss", "sub", "azp", "typ", "preferred_username", "iat", "exp"):
        print("     %-20s %s" % (campo, dados.get(campo)))
    print("     %-20s %s" % ("realm_access.roles", dados.get("realm_access", {}).get("roles")))
    print("     %-20s %s" % ("resource_access", dados.get("resource_access", {})
                             or "{}   <- vazio, e e por isso que ninguem le daqui"))
    print("     %-20s %s segundos" % ("exp - iat", dados["exp"] - dados["iat"]))

    print("\n   ASSINATURA (%d caracteres, base64url, NAO e criptografia:" % len(assinatura))
    print("   qualquer um le o conteudo acima; o que ninguem consegue e altera-lo")
    print("   sem invalidar estes bytes)")
    print("   %s..." % assinatura[:64])

    print("\n5) Agora sim, o curl. O token expira em %s segundos." % resposta["expires_in"])
    print("   export TOKEN='%s'" % acesso)
    print("   curl -i -H \"Authorization: Bearer $TOKEN\" http://localhost:8080/api/v1/pedidos")


if __name__ == "__main__":
    try:
        principal()
    except KeyboardInterrupt:
        sys.exit(130)
